lib/Migration/Version021200Date20240820000001.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\Collectives\Migration;
+
+use Closure;
+use OCA\Collectives\Model\PageInfo;
+use OCA\Collectives\Service\CircleHelper;
+use OCA\Collectives\Service\PageService;
+use OCA\Collectives\Service\SlugGeneratorService;
+use OCP\DB\ISchemaWrapper;
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\DB\Types;
+use OCP\IDBConnection;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version021200Date20240820000001 extends SimpleMigrationStep {
+	private bool $runSlugGeneration = false;
+
+	public function __construct(
+		private IDBConnection $connection,
+		private CircleHelper $circleHelper,
+		private PageService $service,
+		private SlugGeneratorService $slugGeneratorService,
+	) {
+	}
+
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$table = $schema->getTable('collectives_pages');
+		if (!$table->hasColumn('slug')) {
+			$this->runSlugGeneration = true;
+			$table->addColumn('slug', Types::STRING, [
+				'notnull' => false,
+				'default' => false,
+				'length' => 255,
+			]);
+
+			return $schema;
+		}
+
+		return null;
+	}
+
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
+		if (!$this->runSlugGeneration) {
+			return;
+		}
+
+		$queryCollectives = $this->connection->getQueryBuilder();
+		$queryCollectives->select(['id', 'circle_unique_id'])->from('collectives');
+		$resultCollectives = $queryCollectives->executeQuery();
+
+		$queryPages = $this->connection->getQueryBuilder();
+		$queryPages->select(['id'])
+			->from('collectives_pages');
+		$resultPages = $queryPages->executeQuery();
+
+		$update = $this->connection->getQueryBuilder();
+		$update->update('collectives_pages')
+			->set('slug', $update->createParameter('slug'))
+			->where($update->expr()->eq('file_id', $update->createParameter('file_id')));
+
+		while ($rowCollective = $resultCollectives->fetch()) {
+			$circle = $this->circleHelper->getCircle($rowCollective['circle_unique_id'], null, true);
+			/** @var PageInfo[] $pageInfos */
+			$pageInfos = $this->service->findAll($rowCollective['id'], $circle->getOwner()->getUserId());
```

Comment:
Unfortunately this won't work for trashed pages and neither for trashed collectives. I guess it's cleaner to write a dedicated function that unconditionally gets the filenames for all pages from `oc_collectives_pages` for a specific collective and iterates over it.

Please make sure you have trashed collectives *and* active collectives with trashed pages when testing it :wink:

Reply:
I will double check with trashed collectives/pages but IMHO this is fine™. Code should work properly even if slugs are not generated.

Reply:
Well, with a collective in the collectives trash, the commented line throws a `NotFoundException` for me. This can be mitigated by adjusting the query to fetch collectives to only take into account non-trashed collectives:

```diff
--- a/lib/Migration/Version021500Date20240820000001.php
+++ b/lib/Migration/Version021500Date20240820000001.php
@@ -57,7 +57,9 @@ class Version021500Date20240820000001 extends SimpleMigrationStep {
                }
 
                $queryCollectives = $this->connection->getQueryBuilder();
-               $queryCollectives->select(['id', 'circle_unique_id'])->from('collectives');
+               $queryCollectives->select(['id', 'circle_unique_id'])
+                       ->from('collectives')
+                       ->where('trash_timestamp IS NULL');
                $resultCollectives = $queryCollectives->executeQuery();
 
                $queryPages = $this->connection->getQueryBuilder();
```

But still the commented line breaks in my tests with `NotFoundException` as the collective folder cannot be accessed - at least when I run the migration by calling `occ upgrade` on the commandline. I would have to further look into it, but I guess it's because the occ command doesn't run as a specific user, so the collective mountpoints are not accessible.

Could you check if the migration works smoothly if you have different users with collectives and pages and if you trigger the migration via running `occ upgrade`?

You can easily test the migrations again after the following SQL statements:

```sql
ALTER TABLE oc_collectives DROP COLUMN `slug`;
ALTER TABLE oc_collectives_pages DROP COLUMN `slug`;
DELETE FROM oc_migrations WHERE app='collectives' AND version='021500Date20240820000000';
DELETE FROM oc_migrations WHERE app='collectives' AND version='021500Date20240820000001';
```

Since the migration versions changed now, you might have to adjust this.

Afterwards just change the app version number in `appinfo/info.xml` (doesn't matter if you bump or lower it). This will make Nextcloud run the upgrade steps which will run all migrations that are not registered as already executed.

If I do this with pages in the trash, the commented line will throw a `NotFoundException`. This can be mitigated by the following code:

```php
try {
	$pageInfos = $this->service->findAll($rowCollective['id'], $circle->getOwner()->getUserId());
	} catch (NotFoundException
```

Reply:
I was able to reproduce an issue. But it is very strange :confused: .

So, for some reason `PageService::findAll` throws an exception during upgrade if at least 1 page in collective was trashed. But same method works fine in API.


```
string(29) "/admin/files/Collectives/aaaa"  <---- exception message with path + collective name

#0 /var/www/html/lib/private/Files/Node/Folder.php(107): OC\Files\Node\Root->get()
#1 /var/www/html/apps/collectives/lib/Service/PageService.php(100): OC\Files\Node\Folder->get() <--- why collective folder not available? Only 1 page was trashed, not whole collective
#2 /var/www/html/apps/collectives/lib/Service/PageService.php(498): OCA\Collectives\Service\PageService->getCollectiveFolder()
#3 /var/www/html/apps/collectives/lib/Migration/Version021500Date20240820000001.php(78): OCA\Collectives\Service\PageService->findAll()
#4 /var/www/html/lib/private/DB/MigrationService.php(522): OCA\Collectives\Migration\Version021500Date20240820000001->postSchemaChange()
#5 /var/www/html/lib/private/DB/MigrationService.php(395): OC\DB\MigrationService->executeStep()
#6 /var/www/html/lib/private/legacy/OC_App.php(694): OC\DB\MigrationService->migrate()
#7 /var/www/html/lib/private/Updater.php(325): OC_App::updateApp()
#8 /var/www/html/lib/private/Updater.php(236): OC\Updater->doAppUpgrade()
#9 /var/www/html/lib/private/Updater.php(100): OC\Updater->doUpgrade()
#10 /var/www/html/core/Command/Upgrade.php(190): OC\Updater->upgrade()
#11 /var/www/html/3rdparty/symfony/console/Command/Command.php(326): OC\Core\Command\Upgrade->execute()
#12 /var/www/html/3rdparty/symfony/console/Application.php(1078): Symfony\Component\Console\Command\Command->run()
#13 /var/www/html/3rdparty/symfony/console/Application.php(324): Symfony\Component\Console\Application->doRunCommand()
#14 /var/www/html/3rdparty/symfony/console/Application.php(175): Symfony\Component\Console\Application->doRun()
#15 /var/www/html/lib/private/Console/Application.php(187): Symfony\Component\Console\Application->run()
#16 /var/www/html/console.php(87): OC\Console\Application->run()
#17 /var/www/html/occ(11): require_once('...')
#18 {main}
```